### PR TITLE
Fix: -Wextra warnings

### PR DESCRIPTION
**Pull request by zfields**
Fix warnings generated from `-Wextra -Wno-unused-parameter` compiler flags.

Related to issue #30914

### CODE REVIEW

**Comment by dcpleung** [comment]
Please fix the compliance check failures, and follow the contribution guidelines on the required formats.

**Comment by zfields** [comment]
Now what do we do?

**Review by jfischer-no — CHANGES_REQUESTED** [review]
Commit message is missing proper title and motivation/explanation.
"Related to GH Issue https://github.com/zephyrproject-rtos/zephyr/issues/30914" does not belong to the commit message, git does not care about github. Readers of git history should not be forced to go somewhere to get information.

**Comment by zfields** [comment]
My apologies, I simply  intended for it to be a pointer to additional context.

Would the following work better?

> **warnings: clear compiler warnings**
> Fix warnings generated from -Wextra -Wno-unused-parameter compiler flags.
> 
> Signed-off-by: Zachary J. Fields <zachary_fields@yahoo.com>

Here is a recent example I'm basing this from:

> **samples: cdc_acm: enable optional new USB device support** 
> Add code and configuration to enable new USB device support.
>
> Signed-off-by: Johann Fischer <johann.fischer@nordicsemi.no>

**Comment by zfields** [comment]
What do we need to do to move this forward?

The person who has requested changes, has only requested changes in the commit message, which appears to be out-of-line with the "Contribution Guidelines". I have asked for clarification, but I have received none.

The intent of this PR is to fix compiler warnings, which aren't necessarily bugs and shouldn't really exist. Perhaps someone could review the actual work to ensure my solution meets the author's original intent (it's my best guess based on reading the code contextually).

Without the compiler errors resolved, it makes it more difficult to create robust Zephyr applications, so this is a stumbling block for adoption of the platform. Please help me get this through.

File: arch/arm/core/aarch32/cortex_m/fault.c [hunk 1]
```diff
@@ -900,7 +900,7 @@ static inline z_arch_esf_t *get_esf(uint32_t msp, uint32_t psp, uint32_t exc_ret
 	*nested_exc = false;
 
 	if ((exc_return & EXC_RETURN_INDICATOR_PREFIX) !=
-			EXC_RETURN_INDICATOR_PREFIX) {
+			(uint32_t)EXC_RETURN_INDICATOR_PREFIX) {
```
Comment (stephanosio) [inline arch/arm/core/aarch32/cortex_m/fault.c:903]: The correct solution here would be to make `EXC_RETURN_INDICATOR_PREFIX` unsigned:

```c
#define EXC_RETURN_INDICATOR_PREFIX     (0xFFU << 24)
```

File: drivers/interrupt_controller/intc_exti_stm32.c [hunk 2]
```diff
@@ -113,7 +113,7 @@ void stm32_exti_enable(int line)
 {
 	int irqnum = 0;
 
-	if (line >= ARRAY_SIZE(exti_irq_table)) {
+	if (line >= (int)ARRAY_SIZE(exti_irq_table)) {
```
Comment (stephanosio) [inline drivers/interrupt_controller/intc_exti_stm32.c:116]: `line` should never be a negative number, so its type should be `unsigned int`.

File: drivers/timer/cortex_m_systick.c [hunk 3]
```diff
@@ -201,8 +201,8 @@ void sys_clock_set_timeout(int32_t ticks, bool idle)
 		delay =
 		 ((delay + CYC_PER_TICK - 1) / CYC_PER_TICK) * CYC_PER_TICK;
 		delay -= unannounced;
-		delay = MAX(delay, MIN_DELAY);
-		if (delay > MAX_CYCLES) {
+		delay = MAX(delay, (uint32_t)MIN_DELAY);
+		if (delay > (uint32_t)MAX_CYCLES) {
```
Comment (stephanosio) [inline drivers/timer/cortex_m_systick.c:205]: same here. the definition itself ought to be unsigned.

File: subsys/usb/device/class/cdc_acm.c [hunk 4]
```diff
@@ -278,7 +278,7 @@ static void cdc_acm_read_cb(uint8_t ep, int size, void *priv)
 	}
 
 	wrote = ring_buf_put(dev_data->rx_ringbuf, dev_data->rx_buf, size);
-	if (wrote < size) {
+	if ((int)wrote >= 0 && (int)wrote < size) {
```
Comment (stephanosio) [inline subsys/usb/device/class/cdc_acm.c:281]:
```suggestion
	if (wrote >= 0U && (int)wrote < size) {
```

**Review by stephanosio — CHANGES_REQUESTED** [review]

File: drivers/interrupt_controller/intc_exti_stm32.c (hunk 2 above)
Comment (zfields) [reply drivers/interrupt_controller/intc_exti_stm32.c:116]: That was my initial inclination, but I was afraid to adjust an API without consulting with a `Member` first.

It appears the only reference to this API is in `drivers/gpio/gpio_stm32.c`, and it passes in a `uint8_t` as a parameter. It should also maintain ABI compatibility, because `unsigned int` and `int` are the same size. Therefore, this seems like a low risk modification to the API.

File: drivers/timer/cortex_m_systick.c (hunk 3 above)
Comment (zfields) [reply drivers/timer/cortex_m_systick.c:205]: There are cascading effects of changing this MACRO to be unsigned.

```
/home/blues/zephyrproject/zephyr/drivers/timer/cortex_m_systick.c:173:58: warning: operand of '?:' changes signedness from 'int32_t' {aka 'int'} to 'unsigned int' due to unsignedness of other operand [-Wsign-compare]
  173 |         ticks = (ticks == K_TICKS_FOREVER) ? MAX_TICKS : ticks;
      |                                                          ^~~~~
```

Where `ticks` is a parameter in the current function:
```
void sys_clock_set_timeout(int32_t ticks, bool idle)
{
```

Then we would need to change the function definition to become `uint32_t`.

However, this seems to be okay, as the function does contain a `CLAMP` macro that appears to force the value to be greater than or equal to zero. I'll go ahead and update the full scope of the cascading effects, and you will be able to see if you like it in the next commit.

Comment (zfields) [reply drivers/timer/cortex_m_systick.c:205]: Upon further investigation, the documentation appears to be wrong (or misleading).

Specifically...

> Similarly a ticks value of zero (or even negative) is legal and
> treated identically: it simply indicates the kernel would like the
> next tick announcement as soon as possible.

**`drivers/timer/cortex_m_systick.c:162`**
```
if (IS_ENABLED(CONFIG_TICKLESS_KERNEL) && idle && ticks == K_TICKS_FOREVER) {
```

It seems like the comparison should be whether or not `ticks <= 0`, because `K_TICKS_FOREVER` is defined as `-1`.

**`include/zephyr/sys_clock.h:45-51`**
```
#ifdef CONFIG_TIMEOUT_64BIT
typedef int64_t k_ticks_t;
#else
typedef uint32_t k_ticks_t;
#endif

#define K_TICKS_FOREVER ((k_ticks_t) -1)
```

Later in the code, they have attempted to compensate for any negative number by performing the following transformation.

**`drivers/timer/cortex_m_systick.c:173-174`**
```
	ticks = (ticks == K_TICKS_FOREVER) ? MAX_TICKS : ticks;
	ticks = CLAMP(ticks - 1, 0, (int32_t)MAX_TICKS);
```

However, this results in `-1` (`K_TICKS_FOREVER`) being treated differently than -2, which will become zero.

This definitely appears to be a bug, but I don't know how it should be handled. :shrug:

Comment (zfields) [reply drivers/timer/cortex_m_systick.c:205]: I will need input from the team before I know how it should be handled. :shrug:

**Comment by jfischer-no** [comment]
> What do we need to do to move this forward?

Again,
> Commit message is missing proper title and motivation/explanation.

I start review by reading commit message, and it says nothing at all about your motivation.
`-Wextra` enables multiple extra warning flag**s**.  Do they all make sense for the project and why? Are they all or for a specific build configuration? Logical would be to split PR into several commits, either by code area or specific flag enabled by `-Wextra`, maybe it will help you to form a proper commit message. 

> The person who has requested changes, has only requested changes in the commit message, which appears to be out-of-line with the "Contribution Guidelines". I have asked for clarification, but I have received none.
> 

Your entire commit message consists of ridiculous "Related to GH Issue #30914". I would not even accept it if it was written like that in "Contribution Guidelines".

File: drivers/timer/cortex_m_systick.c (hunk 3 above)
Comment (zfields) [reply drivers/timer/cortex_m_systick.c:205]: After stepping away for a bit, I realized I can move forward and leave the exact same problematic logic in place.

I just need to change the result of the CLAMP macro to an unsigned variable, and propagate the variable internally throughout the function.

**Comment by zfields** [comment]
@jfischer-no, although you are being condescending (e.g. using the word "ridiculous"), I do appreciate you providing actionable context so that I may improve the commit message.

I apologize for the original terse commit message, but the commit title seemed self-explanatory to me. Typically warnings are frowned upon and understood to be oversights or omissions. I thought the problem was summarized nicely in [GHI #30914](https://github.com/zephyrproject-rtos/zephyr/issues/30914), which is why I linked out to it.

I'm happy to paste the issue text in the commit message, if that works for you?

> When compiling with the above option, I want to see no warning/errors at all related to Zephyr directory. This will make Zephyr code more robust, and I will be able to use this option by default for my project. Right now, since Zephyr directory produces a ot of warnings, it is hard to use it for my application, as I have to search through hundred lines of compiler output.

Honestly, I'm still not sure what you're looking for, or what more I could explain. Perhaps you could offer an example of a good commit message?

File: drivers/timer/cortex_m_systick.c (hunk 3 above)
Comment (zfields) [reply drivers/timer/cortex_m_systick.c:205]: It's not as easy to bypass as I'd hoped.

When the value of `MAX_TICKS` is sufficiently large, then the outcome of the logic in the following two lines...

```cpp
	ticks = (ticks == K_TICKS_FOREVER) ? MAX_TICKS : ticks;
	ticks = CLAMP(ticks - 1, 0, (int32_t)MAX_TICKS);
```

... is heavily dependent on the "signedness" of the integer. Which is why I expect the bounding logic looks so bizarre.

In plain English, I would suspect the intent of the function is to preserve the special meaning of `K_TICKS_FOREVER`, otherwise bound `ticks` between `0` and `MAX_TICKS`.

Let's look at following scenarios, remembering `ticks` is a signed value, `K_TICKS_FOREVER` is -1, and pretending that `k_ticks_t` is a `uint8_t` for ease of reading...

1. **ticks == 75 and MAX_TICKS == 100**
   1. `ticks` remains 75  // ticks = (ticks == K_TICKS_FOREVER) ? MAX_TICKS : ticks;
   2. `ticks` becomes 74  // ticks = CLAMP(ticks - 1, 0, (int32_t)MAX_TICKS);
2. **ticks == 127 and MAX_TICKS == 100**
   1. `ticks` remains 127  // ticks = (ticks == K_TICKS_FOREVER) ? MAX_TICKS : ticks;
   2. `ticks` becomes 100  // ticks = CLAMP(ticks - 1, 0, (int32_t)MAX_TICKS);
3. **ticks == -1 and MAX_TICKS == 100**
   1. `ticks` becomes 100  // ticks = (ticks == K_TICKS_FOREVER) ? MAX_TICKS : ticks;
   2. `ticks` becomes 99  // ticks = CLAMP(ticks - 1, 0, (int32_t)MAX_TICKS);
4. **ticks == -1 and MAX_TICKS == 200**
   1. `ticks` becomes -56 (i.e. 200)  // ticks = (ticks == K_TICKS_FOREVER) ? MAX_TICKS : ticks;
   2. `ticks` becomes 0  // ticks = CLAMP(ticks - 1, 0, (int32_t)MAX_TICKS);
5. **ticks == -2 and MAX_TICKS == 200**
   1. `ticks` remains -2  // ticks = (ticks == K_TICKS_FOREVER) ? MAX_TICKS : ticks;
   2. `ticks` becomes 0  // ticks = CLAMP(ticks - 1, 0, (int32_t)MAX_TICKS);
6. **ticks == 100 and MAX_TICKS == 200**
   1. `ticks` remains 100  // ticks = (ticks == K_TICKS_FOREVER) ? MAX_TICKS : ticks;
   2. `ticks` becomes -56  // ticks = CLAMP(ticks - 1, 0, (int32_t)MAX_TICKS);

As you can see, this logic leads to some pretty wild and unexpected outcomes.

Comment (zfields) [reply drivers/timer/cortex_m_systick.c:205]: @stephanosio, I could really use your guidance here.
